### PR TITLE
Add h-refinement criterion based on power in highest modes

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Similar to the Persson TCI.

Figure 8

Bottom panel are the modes in each element. Top right axis and horizontal lines in top panel show the smoothness indicator.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by nilsvu** [comment]
@nilsdeppe @yoonso0-0 here's the Persson-based AMR criterion. Quite simple implementation. Pinging you because we talked about this, feel free to review if you want to.

File: src/ParallelAlgorithms/Amr/Criteria/Persson.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  auto matrices = make_array<Dim>(std::cref(identity));
+  gsl::at(matrices, dimension) = Spectral::filtering::zero_lowest_modes(
+      mesh.slice_through(dimension),
+      mesh.extents(dimension) - num_highest_modes);
+  set_number_of_grid_points(filtered_component_buffer,
+                            mesh.number_of_grid_points());
+  apply_matrices(filtered_component_buffer, matrices, tensor_component,
+                 mesh.extents());
+  // Take the L2 norm over all grid points
+  return sqrt(alg::accumulate(square(*filtered_component_buffer), 0.) /
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Persson.cpp]: I'm worried about expressions and template parameters here. Why not use Blaze L2 norm?

File: src/ParallelAlgorithms/Amr/Criteria/Persson.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+void max_over_components(const gsl::not_null<std::array<Flag, Dim>*> result,
+                         const gsl::not_null<DataVector*> buffer,
+                         const DataVector& tensor_component,
+                         const Mesh<Dim>& mesh, const size_t num_highest_modes,
+                         const double relative_tolerance,
+                         const double absolute_tolerance,
+                         const double coarsening_factor) {
+  for (size_t d = 0; d < Dim; ++d) {
+    // Skip this dimension if we have already decided to refine it
+    if (gsl::at(*result, d) == Flag::IncreaseResolution) {
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Persson.cpp]: Is `Split` not higher than `IncreaseResolution`?

File: src/ParallelAlgorithms/Amr/Criteria/Persson.hpp [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+/// @{
+/*!
+ * \brief Computes an anisotropic smoothness indicator based on the power in the
+ * highest modes
+ *
+ * This smoothness indicator is the L2 norm of the tensor component with the
+ * lowest N - M modes filtered out, where N is the number of grid points in the
+ * given dimension and M is `num_highest_modes`.
+ *
+ * This strategy is similar to the Persson troubled-cell indicator (see
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Persson.hpp:43]: Can you add some math here? This is similar to Persson in the sense of it uses spectral coefficients, but you have completely different parameters from Persson, so you really need to document this. I don't understand what value I should choose when compare to the alpha value of the Persson TCI. That needs to be made extremely clear because that value is central to the Persson TCI.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/ParallelAlgorithms/Amr/Criteria/Persson.cpp (hunk 1 above)
Comment (nilsvu) [reply src/ParallelAlgorithms/Amr/Criteria/Persson.cpp]: Blaze's L2 norm doesn't divide by the number of grid points, and I found the docs confusing. The docs say "computes the _squared_ L1/2/3/4 norm" (https://bitbucket.org/blaze-lib/blaze/wiki/Vector%20Operations#!norms). I found it clearer to do this explicit sum-divide-sqrt.

File: src/ParallelAlgorithms/Amr/Criteria/Persson.cpp (hunk 2 above)
Comment (nilsvu) [reply src/ParallelAlgorithms/Amr/Criteria/Persson.cpp]: Ohh good catch, this should be `Split`.

File: src/ParallelAlgorithms/Amr/Criteria/Persson.cpp (hunk 1 above)
Comment (nilsdeppe) [reply src/ParallelAlgorithms/Amr/Criteria/Persson.cpp]: Okay, but the performance here is unclear. Just use the Blaze one.

Comment (nilsvu) [reply src/ParallelAlgorithms/Amr/Criteria/Persson.cpp]: You mean `sqrt(square(blaze::l2Norm(*buffer)) / num_points)`?

Comment (nilsdeppe) [reply src/ParallelAlgorithms/Amr/Criteria/Persson.cpp]: Why not `blaze::l2Norm(*buffer) / sqrt(static_cast<double>(num_points));`?

Comment (nilsdeppe) [reply src/ParallelAlgorithms/Amr/Criteria/Persson.cpp]: or maybe: `(1.0 / sqrt(static_cast<double>(num_points))) * blaze::l2Norm(*buffer)`. :shrug:

Comment (nilsvu) [reply src/ParallelAlgorithms/Amr/Criteria/Persson.cpp]: Ok

File: src/ParallelAlgorithms/Amr/Criteria/Persson.hpp (hunk 3 above)
Comment (nilsvu) [reply src/ParallelAlgorithms/Amr/Criteria/Persson.hpp:43]: relative tolerance = N^{-\alpha}

I'll just switch to `Alpha` as a parameter.

**Comment by nilsvu** [comment]
Pushed a fixup

**Review by nilsdeppe — DISMISSED** [review]
LGTM, please squash!

**Comment by yoonso0-0** [comment]
I believe @nilsdeppe resolved technical parts, so just a small suggestion. What I see from here is that technically this is slightly different from the exact implementation of the Persson TCI, in that this AMR criteria uses `max(U)`, highest modes norm divided by `sqrt(N)` etc. So I think it'd be better to rename this criteria to something other than the term `Persson` to avoid potential confusion.

**Comment by yoonso0-0** [comment]
From the SIAM conference I saw many people using the term "spectral shock sensor" or "modal sensor", I would pick something like `SpectralSensor` or `ModalSensor`. Welcome any opinions from Nils and Larry :)

File: src/ParallelAlgorithms/Amr/Criteria/TruncationError.cpp [hunk 4]
```diff
@@ -29,6 +29,7 @@ void max_over_components(
   // still satisfy the target with the highest mode removed will the element
   // decrease p refinement in that dimension.
   PowerMonitors::power_monitors(power_monitors_buffer, tensor_component, mesh);
+  const double umax = max(abs(tensor_component));
```
Comment (kidder) [inline src/ParallelAlgorithms/Amr/Criteria/TruncationError.cpp:32]: this change should be in a separate commit

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsvu** [comment]
@yoonso0-0 I think this is similar enough to the Persson TCI that sharing the name helps transfer some of the experience from the TCI to this AMR criterion. No strong opinion though.

**Comment by nilsdeppe** [comment]
I personally think it's close enough to the Persson TCI (and might even be better in general) that the name is fine :shrug: we can rename later if need be :)

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]
